File: 3sem/4lab/Banks/BankAccounts/DepositBankAccount.cs
using Banks.Clocks;
using Banks.Entities;
using Banks.Exceptions;
using Banks.Models.BankConfigurations;

namespace Banks.BankAccounts;

public class DepositBankAccount : IBankAccount
{
    private readonly IClock _clock;
    private decimal _dailyPercents;

    public DepositBankAccount(decimal deposit, Bank bank, Client client, IClock clock)
    {
        ArgumentNullException.ThrowIfNull(bank);
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(clock);

        if (deposit < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        _dailyPercents = 0;

        Balance = deposit;
        Client = client;
        Configuration = bank.Configuration.DepositConfiguration;
        InterestRate = bank.Configuration.DepositConfiguration.Selector.GetInterestRate(deposit);
        EndDate = clock.Now + bank.Configuration.DepositConfiguration.Duration;
        _clock = clock;

        Id = Guid.NewGuid();

        clock.DayChanged += OnDayChanged;
        clock.MonthChanged += OnMonthChanged;
    }

    public Guid Id { get; }
    public Client Client { get; }
    public DepositBankAccountConfiguration Configuration { get; }
    public decimal InterestRate { get; }
    public decimal Balance { get; private set; }
    public DateTime EndDate { get; }

    public bool CanDeposit(decimal moneyAmount)
    {
        return Client.IsVerified() || moneyAmount <= Configuration.UnverifiedLimit;
    }

    public bool CanWithdraw(decimal moneyAmount)
    {
        if (!Client.IsVerified() && moneyAmount > Configuration.UnverifiedLimit)
            return false;

        return _clock.Now >= EndDate && moneyAmount <= Balance;
    }

    public void Deposit(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        Balance += moneyAmount;
    }

    public void Withdraw(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        if (!CanWithdraw(moneyAmount))
            throw InvalidBankAccountOperationException.OnShortageOfMoney(moneyAmount, Balance);

        Balance -= moneyAmount;
    }

    private void OnDayChanged()
    {
        if (_clock.Now < EndDate)
            _dailyPercents += Balance * InterestRate / 365;
        else
            _clock.DayChanged -= OnDayChanged;
    }

    private void OnMonthChanged()
    {
        if (_clock.Now >= EndDate)
            _clock.MonthChanged -= OnMonthChanged;

        Balance += _dailyPercents;
        _dailyPercents = 0;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/BankAccounts/DepositBankAccount.java
package banksRemake.domain.bankAccounts;

import banksRemake.domain.clocks.Clock;
import banksRemake.domain.clocks.ClockSubscriber;
import banksRemake.domain.entities.Bank;
import banksRemake.domain.entities.Client;
import banksRemake.domain.exceptions.InvalidBankAccountOperationException;
import banksRemake.domain.models.bankConfigurations.DepositBankAccountConfiguration;

import java.util.Date;
import java.util.UUID;

/**
 * A bank account, without the possibility of withdrawing for a certain period of time and with an interest rate that depends on the amount.
 */
public class DepositBankAccount implements BankAccount, ClockSubscriber {
    private final UUID id;
    private final Client client;
    private final DepositBankAccountConfiguration configuration;
    private final Clock clock;
    private final double interestRate;
    private double balance;
    private double dailyPercents;
    private final Date endDate;

    public DepositBankAccount(double deposit, Bank bank, Client client, Clock clock) throws InvalidBankAccountOperationException {
        if (deposit < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        this.dailyPercents = 0;

        this.balance = deposit;
        this.client = client;
        this.configuration = bank.getConfiguration().depositBankAccountConfiguration();
        this.interestRate = configuration.selector().getInterestRate(deposit);
        this.endDate = Date.from(clock.getNowDate().toInstant().plus(configuration.duration()));
        this.clock = clock;

        id = UUID.randomUUID();

        clock.addSubscriber(this);
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getBalance() {
        return balance;
    }

    @Override
    public Client getClient() {
        return client;
    }

    @Override
    public boolean canDeposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        return client.isVerified() || moneyAmount <= configuration.unverifiedLimit();
    }

    @Override
    public boolean canWithdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!client.isVerified() && moneyAmount > configuration.unverifiedLimit()) {
            return false;
        }

        return clock.getNowDate().after(endDate) && moneyAmount <= balance;
    }

    @Override
    public void deposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        balance += moneyAmount;
    }

    @Override
    public void withdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!canWithdraw(moneyAmount)) {
            throw InvalidBankAccountOperationException.onShortageOfMoney(moneyAmount, balance);
        }

        balance -= moneyAmount;
    }

    @Override
    public void reactOnDayChanged() {
        if (clock.getNowDate().after(endDate)) {
            dailyPercents += balance * interestRate / 365;
        }
        else {
            clock.removeSubscriber(this);
        }
    }

    @Override
    public void reactOnMonthChanged() {
        if (clock.getNowDate().after(endDate)) {
            clock.removeSubscriber(this);
        }

        balance += dailyPercents;
        dailyPercents = 0;
    }
}
